### DISCUSSION TITLE
FastAPI does not block new requests when waiting for a request.post it called internally

### DISCUSSION THREAD

**Discussion by f-grimaldi** [fastapi/fastapi#12202] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@app.post('/api/v1/some_api', tags=['some_tag'], response_model=PydanticOuput)
def some_funtion(input: PydanticInput, auth: HTTPAuthorizationCredentials = Security(security)) -> JSONResponse:
    request_id = str(uuid.uuid4())
    with app.logger.contextualize(request_id=request_id):
        app.logger.info(f'Request received')
        response = requests.post(OTHER_API, json=input.dict(), timeout=general_config.api_timeout)
        return JSONResponse(response.dict(), status_code=200)
```

### Description

I have this FastAPI program (launched by `nohup ddtrace-run uvicorn main:app --workers 4 &`) that basically acts as an _orchestrator_, calling two different services, built with fastapi as well, on backend (in the code you see only one for semplicity). All these services are defined with `def` and not `async def` because i want to process all the requests sequentially.

Now, the problem is  that, sometimes, Fastapi accepts and start processings a second request when the previous one is still waiting for the response from the `request.post` function and i do not understand why, since as far i understood, the _request_ is blocking.

Here it's a real (slightly modified and simplified) logging COMING FROM ONE SINGLE UVICORN THREAD (each uvicorn worker write into a log file with an unique uuid):
```
INFO     2024-09-11 12:45:32.799 request id: 310fb485-6e4c-496d-8584-3159d736eadd - Request received                                               <- REQUEST #1
INFO     2024-09-11 12:45:32.800 request id: 310fb485-6e4c-496d-8584-3159d736eadd - Calling database to check if answer already exists…            <- REQUEST #1
INFO     2024-09-11 12:45:32.809 request id: 310fb485-6e4c-496d-8584-3159d736eadd - Not found in database                                          <- REQUEST #1
INFO     2024-09-11 12:45:32.825 request id: 310fb485-6e4c-496d-8584-3159d736eadd - Calling backend with requests                                  <- REQUEST #1
INFO     2024-09-11 12:45:33.095 request id: c455dea8-00e7-4eec-9df9-cd821f3f563a - Request received                                               <- REQUEST #2
INFO     2024-09-11 12:45:33.096 request id: c455dea8-00e7-4eec-9df9-cd821f3f563a - Calling database to check if answer already exists…            <- REQUEST #2
INFO     2024-09-11 12:45:33.102 request id: c455dea8-00e7-4eec-9df9-cd821f3f563a - Not found in database                                          <- REQUEST #2
INFO     2024-09-11 12:45:33.130 request id: c455dea8-00e7-4eec-9df9-cd821f3f563a - Calling backend with requests                                  <- REQUEST #2
INFO     2024-09-11 12:45:44.715 request id: c455dea8-00e7-4eec-9df9-cd821f3f563a - Succesfull response                                            <- REQUEST #2
ERROR    2024-09-11 12:45:52.839 request id: 310fb485-6e4c-496d-8584-3159d736eadd - Read timed out. (read timeout=20.0)                            <- REQUEST #1
```

### Operating System

Linux

### Operating System Details

OS coming from dokcerhub image python3:11 bullseye

### FastAPI Version

0.110.0

### Pydantic Version

2.6.4

### Python Version

3.11

### Additional Context

FastAPI webservice deployed via dockerfile in k8s AWS environment

This is the _requirements.txt_ file 
```
ddtrace==2.8.5
fastapi==0.110.0
loguru==0.7.0
numpy==1.26.4
pydantic==2.6.4
pydantic_core==2.16.3
pydantic_settings==2.2.1
pymongo[srv]==3.11.0
requests==2.31.0
uvicorn==0.29.0
```

**Comment by YuriiMotov** [comment, marked as answer]
FastAPI runs sync endpoints in the threadpool (in separate threads), so they don't block the app and it can process other requests in parallel.

To achieve the behavior you want, you need to implement any type of lock mechanism (so that only one endpoint call at a time could acquire the lock and be executed)

**Reply by YuriiMotov** [reply]
Something similar to this behavior could be achieved if you define your endpoints as `async def` and use sync blocking code inside.
But it's bad idea since it will block all your app and may cause the timeout and reloading the worker.

**Reply by f-grimaldi** [reply]
Thanks for the info. It's what i was really looking for to understand.
Now i will think about a possible implementation to fix this behavior, thanks again!
